src/Database/TigerBeetle/Raw/Client.hs

```diff
@@ -74,6 +85,98 @@ data ClientState = ClientState
 
 newtype ClientHandle = ClientHandle { tvar :: TVar ClientState }
 
+data Address = Address { getAddress :: Text }
+  deriving (Eq, Show)
+
+toCString :: Address -> ((Ptr CChar, Int) -> IO a) -> IO a
+toCString = BS.useAsCStringLen . TE.encodeUtf8 . getAddress
+
+data ClientInitError
+  = Unexpected
+  | OutOfMemory
+  | AddressInvalid
+  | AddressLimitExceeded
+  | SystemResources
+  | NetworkSubsystem
+  deriving (Eq, Show)
+
+-- | Create a 'Client' initialization error from a 'TBInitStatus'.
+--
+-- Asserts that @err@ is not 'FFI.Success', throws an exception at
+-- runtime.
+toClientInitError :: TBInitStatus -> ClientInitError
+toClientInitError err = assert (err /= FFI.Success) $
+  case err of
+    FFI.Unexpected           -> Unexpected
+    FFI.OutOfMemory          -> OutOfMemory
+    FFI.AddressInvalid       -> AddressInvalid
+    FFI.AddressLimitExceeded -> AddressLimitExceeded
+    FFI.SystemResources      -> SystemResources
+    FFI.NetworkSubsystem     -> NetworkSubsystem
+    FFI.Success              -> error "toClientInitError: Success is not an error"
+
+type Client = ForeignPtr FFI.TBClient
+
+clientFinalizer :: Ptr TBClient -> IO ()
+clientFinalizer clientPtr = do
+  result <- FFI.tbClientDeinit clientPtr
+  case result of
+    FFI.ClientOk -> pure ()
+    FFI.ClientInvalid ->
+      error $ "tbClientFinalizer (invalid clientPtr): " ++ show clientPtr
```

Comment:
Oh I see, you basically do use `tbClientDeinit` as the finalizer. Maybe this could take a user provided function to run on the result (that way consumers could add their own logging, or do something with the `TBClientStatus`

Reply:
That would be an interesting addition. It's a bit risky in a finalizer to do much other than call the foreign cleanup function as this will be called by the GHC runtime when the last reference to the pointer goes out of scope. Arbitrary `IO` in this case could cause problems if users aren't careful.

Also not sure we should even throw an `IO` exception here since that would block the runtime from getting to the cleanup code and GC on the objects.

A pure function would probably be relatively safe or else maybe an `unsafe` version of this finalizer?

Reply:
@agentultra oh good call, I didn't even think about exceptions. There is probably some clever way we could mask the user action / catch any exceptions. But this is much easier as an initial implementation. Looks good to me!